snake.go

```diff
@@ -1,64 +1,81 @@
-//
-// Fast snake-case implementation.
-//
+// Package snakecase - Super-Fast snake-case implementation.
 package snakecase
 
+const underscoreByte = '_'
+
 // Snakecase the given string.
 func Snakecase(s string) string {
-	b := make([]byte, 0, 64)
-	l := len(s)
-	i := 0
-
-	// loop until we reached the end of the string
-	for i < l {
+	idx := 0
+	hasLower := false
+	hasUnderscore := false
+	lowercaseSinceUnderscore := false
 
-		// skip leading bytes that aren't letters or numbers
-		for i < l && !isWord(s[i]) {
-			i++
+	// loop through all good characters:
+	// - lowercase
+	// - digit
+	// - underscore (as long as the next character is lowercase or digit)
+	for ; idx < len(s); idx++ {
+		if isLower(s[idx]) {
+			hasLower = true
+			if hasUnderscore {
+				lowercaseSinceUnderscore = true
+			}
+			continue
+		} else if isDigit(s[idx]) {
+			continue
+		} else if s[idx] == underscoreByte && idx > 0 && idx < len(s)-1 && (isLower(s[idx+1]) || isDigit(s[idx+1])) {
+			hasUnderscore = true
+			lowercaseSinceUnderscore = false
+			continue
 		}
+		break
+	}
 
-		if i < l && len(b) != 0 {
-			b = append(b, '_')
-		}
+	if idx == len(s) {
+		return s // no changes needed, can just borrow the string
+	}
 
-		// Append all leading uppercase or digits
-		for i < l {
-			if c := s[i]; !isHead(c) {
-				break
-			} else {
-				b = append(b, toLower(c))
-			}
-			i++
+	// if we get here then we must need to manipulate the string
+	b := make([]byte, 0, 64)
```

Comment:
Is 64 still a good choice of size? I used a static size initially because I wanted it to be stack-allocated, but since we know we're gonna have to modify the string and the result will be larger than the original, it's very likely that this buffer will end up being heap allocated.

It may be worth doing `make([]byte, 0, 2 * len(s))` (or some other factor), then instead of `return string(b)` we could do `return *(*string)(unsafe.Pointer(&b))`.
It does add a dependency to `unsafe` which we didn't have before, so possibly increasing the maintenance cost further.

Reply:
@achille-roussel I did try a few variants including `make([]byte, 0, 2 * len(s))` which yielded much poorer performance on my machine. I think it's due the 64 aligning better to the x86_64 arch even though it's larger.

I do like the suggestion of `return *(*string)(unsafe.Pointer(&b))` though, going to try that and see if there is an increase worth the unsafe dependency.

Reply:
here is the result with `make([]byte, 0, 2 * len(s))`:
```
benchmark                                   old ns/op     new ns/op     delta
BenchmarkUnchangedLong-8                    96.0          39.7          -58.65%
BenchmarkUnchangedSimple-8                  41.2          15.2          -63.11%
BenchmarkModifiedUnicode-8                  36.6          59.0          +61.20%
BenchmarkModifiedLong-8                     86.7          118           +36.10%
BenchmarkModifiedLongSpecialChars-8         135           161           +19.26%
BenchmarkModifiedSimple-8                   43.7          62.1          +42.11%
BenchmarkModifiedUnicode2-8                 73.0          110           +50.68%
BenchmarkLeadingUnderscoresDigitUpper-8     39.2          60.9          +55.36%
BenchmarkDigitUpper-8                       34.4          42.5          +23.55%
BenchmarkDigitUpper2-8                      85.2          114           +33.80%
```

so `b` when using `b := make([]byte, 0, 64)` doesn't actually escape to the heap because it never leaves the function nor does it cross any package boundaries:
`go build -gcflags '-m'`
```
./snake.go:85:6: can inline isLower
./snake.go:89:6: can inline isDigit
./snake.go:81:6: can inline isUpper
./snake.go:77:6: can inline isAlphanumeric
./snake.go:78:16: inlining call to isLower
./snake.go:78:30: inlining call to isUpper
./snake.go:78:44: inlining call to isDigit
./snake.go:18:13: inlining call to isLower
./snake.go:24:20: inlining call to isDigit
./snake.go:26:78: inlining call to isLower
./snake.go:26:99: inlining call to isDigit
./snake.go:42:12: inlining call to isUpper
./snake.go:43:31: inlining call to isUpper
./snake.go:43:50: inlining call to isDigit
./snake.go:48:31: inlining call to isLower
./snake.go:48:50: inlining call to isDigit
./snake.go:55:21: inlining call to isAlphanumeric
./snake.go:55:21: inlining call to isLower
./snake.go:55:21: inlining call to isUpper
./snake.go:55:21: inlining call to isDigit
./snake.go:64:31: inlining call to isUpper
./snake.go:64:50: inlining call to isDigit
./snake.go:69:31: inlining call to isLower
./snake.go:69:50: inlining call to isDigit
./snake.go:7:16: leaking param: s to result ~r1 level=0
./snake.go:74:15: string(b) escapes to heap
./snake.go:39:11: Snakecase make([]byte, 0, 64) does not escape
./snake.go:44:37: index bounds check elided
./snake.go:65:37: index bounds check elided
```

but using the unsafe.Pointer causes it to be heap allocated and adversely affects performance:
```
./snake.go:88:6: can inline isLower
./snake.go:92:6: can inline isDigit
./snake.go:84:6: can inline isUpper
./snake.go:80:6: can inline isAlphanumeric
./snake.go:81:16: inlining call to isLower
./snake.go:81:30: inlining call to isUpper
./snake.go:81:44: inlining call to isDigit
./snake.go:20:13: inlining call to isLower
./snake.go:26:20: inlining call to isDigit
./snake.go:28:78: inlining call to isLower
./snake.go:28:99: inlining call to isDigit
./snake.go:44:12: inlining call to isUpper
./snake.go:45:31: inlining call to isUpper
./snake.go:45:50: inlining call to isDigit
./snake.go:50:31: inlining call to isLower
./snake.go:50:50: inlining call to isDigit
./snake.go:57:21: inlining call to isAlphanumeric
./snake.go:57:21: inlining call to isLower
./snake.go:57:21: inlining call to isUpper
./snake.go:57:21: inlining call to isDigit
./snake.go:66:31: inlining call to isUpper
./snake.go:66:50: inlining call to isDigit
./snake.go:71:31: inlining call to isLower
./snake.go:71:50: inlining call to isDigit
./snake.go:9:16: leaking param: s to result ~r1 level=0
./snake.go:41:11: make([]byte, 0, 64) escapes to heap
./snake.go:76:35: Snakecase &b does not escape
./snake.go:46:37: index bounds check elided
./snake.go:67:37: index bounds check elided
```
```
benchmark                                   old ns/op     new ns/op     delta
BenchmarkUnchangedLong-8                    96.0          39.4          -58.96%
BenchmarkUnchangedSimple-8                  41.2          14.7          -64.32%
BenchmarkModifiedUnicode-8                  36.6          49.0          +33.88%
BenchmarkModifiedLong-8                     86.7          88.4          +1.96%
BenchmarkModifiedLongSpecialChars-8         135           133           -1.48%
BenchmarkModifiedSimple-8                   43.7          48.9          +11.90%
BenchmarkModifiedUnicode2-8                 73.0          86.6          +18.63%
BenchmarkLeadingUnderscoresDigitUpper-8     39.2          47.5          +21.17%
BenchmarkDigitUpper-8                       34.4          40.4          +17.44%
BenchmarkDigitUpper2-8                      85.2          88.9          +4.34%

benchmark                                   old MB/s     new MB/s     speedup
BenchmarkUnchangedLong-8                    343.66       838.21       2.44x
BenchmarkUnchangedSimple-8                  266.95       748.43       2.80x
BenchmarkModifiedUnicode-8                  245.89       183.64       0.75x
BenchmarkModifiedLong-8                     334.32       328.22       0.98x
BenchmarkModifiedLongSpecialChars-8         222.20       224.13       1.01x
BenchmarkModifiedSimple-8                   251.79       225.08       0.89x
BenchmarkModifiedUnicode2-8                 355.94       300.10       0.84x
BenchmarkLeadingUnderscoresDigitUpper-8     153.03       126.21       0.82x
BenchmarkDigitUpper-8                       145.42       123.74       0.85x
BenchmarkDigitUpper2-8                      199.58       191.24       0.96x

benchmark                                   old allocs     new allocs     delta
BenchmarkUnchangedLong-8                    1              0              -100.00%
BenchmarkUnchangedSimple-8                  1              0              -100.00%
BenchmarkModifiedUnicode-8                  1              1              +0.00%
BenchmarkModifiedLong-8                     1              1              +0.00%
BenchmarkModifiedLongSpecialChars-8         1              1              +0.00%
BenchmarkModifiedSimple-8                   1              1              +0.00%
BenchmarkModifiedUnicode2-8                 1              1              +0.00%
BenchmarkLeadingUnderscoresDigitUpper-8     1              1              +0.00%
BenchmarkDigitUpper-8                       1              1              +0.00%
BenchmarkDigitUpper2-8                      1              1              +0.00%

benchmark                                   old bytes     new bytes     delta
BenchmarkUnchangedLong-8                    48            0             -100.00%
BenchmarkUnchangedSimple-8                  16            0             -100.00%
BenchmarkModifiedUnicode-8                  3             64            +2033.33%
BenchmarkModifiedLong-8                     48            64            +33.33%
BenchmarkModifiedLongSpecialChars-8         32            64            +100.00%
BenchmarkModifiedSimple-8                   16            64            +300.00%
BenchmarkModifiedUnicode2-8                 3             64            +2033.33%
BenchmarkLeadingUnderscoresDigitUpper-8     5             64            +1180.00%
BenchmarkDigitUpper-8                       5             64            +1180.00%
BenchmarkDigitUpper2-8                      32            64            +100.00%
```

it looks like the way it is is the most efficient.

Reply:
Nice! 👍